src/backends/policy_enforcement/souffle/value.h

```diff
@@ -0,0 +1,124 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+
+#ifndef SRC_BACKENDS_POLICY_ENFORCEMENT_SOUFFLE_VALUE_H_
+#define SRC_BACKENDS_POLICY_ENFORCEMENT_SOUFFLE_VALUE_H_
+
+#include <string>
+#include <tuple>
+#include <variant>
+
+#include "absl/strings/str_format.h"
+#include "absl/strings/str_join.h"
+#include "absl/strings/string_view.h"
+
+namespace raksha::backends::policy_enforcement::souffle {
+
+// The common supertype of any "fully fledged" Souffle value type. Anything
+// descending form `Value` should be able to be used in a `.type` declaration.
+// By this logic, `number`, `symbol`, any record, and any ADT are `Value`s, but
+// individual branches of an ADT are not.
+class Value {
+ public:
+  virtual std::string ToDatalogString() const = 0;
+};
+
+// This is the common supertype of any Souffle value type that can
+// participate in an ADT. This is all Values except Records. Values that can
+// participate in an ADT must have a virtual destructor.
+class AdtMemberValue : public Value {
+ public:
+  virtual ~AdtMemberValue() {}
```

Comment:
I was using the split between `AdtMemberValue` and `Value` to prevent records from being used as members of an ADT branch. I believed them to be illegal there, but looking at the Souffle documentation, I cannot find evidence for that. I'll combine them as you suggest.

Reply:
Fixed.